lib/kube/proxy/forwarder.go

```diff
@@ -1011,51 +1017,143 @@ func (f *Forwarder) execNonInteractive(ctx *authContext, w http.ResponseWriter,
 	}
 
 	eventPodMeta := request.eventPodMeta(request.context, sess.creds)
-	event := &apievents.Exec{
+
+	sessionStart := f.cfg.Clock.Now().UTC()
+
+	serverMetadata := apievents.ServerMetadata{
+		ServerID:        f.cfg.ServerID,
+		ServerNamespace: f.cfg.Namespace,
+		ServerHostname:  sess.teleportCluster.name,
+		ServerAddr:      sess.kubeAddress,
+	}
+
+	sessionMetadata := apievents.SessionMetadata{
+		SessionID: uuid.NewString(),
+		WithMFA:   ctx.Identity.GetIdentity().MFAVerified,
+	}
+
+	connectionMetdata := apievents.ConnectionMetadata{
+		RemoteAddr: req.RemoteAddr,
+		LocalAddr:  sess.kubeAddress,
+		Protocol:   events.EventProtocolKube,
+	}
+
+	sessionStartEvent := &apievents.SessionStart{
 		Metadata: apievents.Metadata{
-			Type:        events.ExecEvent,
+			Type:        events.SessionStartEvent,
+			Code:        events.SessionStartCode,
 			ClusterName: f.cfg.ClusterName,
 		},
-		ServerMetadata: apievents.ServerMetadata{
-			ServerID:        f.cfg.ServerID,
-			ServerNamespace: f.cfg.Namespace,
-		},
-		SessionMetadata: apievents.SessionMetadata{
-			SessionID: uuid.NewString(),
-			WithMFA:   ctx.Identity.GetIdentity().MFAVerified,
-		},
-		UserMetadata: ctx.eventUserMeta(),
-		ConnectionMetadata: apievents.ConnectionMetadata{
-			RemoteAddr: req.RemoteAddr,
-			LocalAddr:  sess.kubeAddress,
-			Protocol:   events.EventProtocolKube,
+		ServerMetadata:            serverMetadata,
+		SessionMetadata:           sessionMetadata,
+		UserMetadata:              ctx.eventUserMeta(),
+		ConnectionMetadata:        connectionMetdata,
+		KubernetesClusterMetadata: ctx.eventClusterMeta(),
+		KubernetesPodMetadata:     eventPodMeta,
+
+		InitialCommand:   request.cmd,
+		SessionRecording: ctx.recordingConfig.GetMode(),
+	}
+
+	sessionStartEvent.Impersonator = ctx.Identity.GetIdentity().Impersonator
+
+	if err := f.cfg.StreamEmitter.EmitAuditEvent(f.ctx, sessionStartEvent); err != nil {
+		f.log.WithError(err).Warn("Failed to emit event.")
+	}
+
+	execEvent := &apievents.Exec{
+		Metadata: apievents.Metadata{
+			Type:        events.ExecEvent,
+			ClusterName: f.cfg.ClusterName,
 		},
+		ServerMetadata:     serverMetadata,
+		SessionMetadata:    sessionMetadata,
+		UserMetadata:       ctx.eventUserMeta(),
+		ConnectionMetadata: connectionMetdata,
 		CommandMetadata: apievents.CommandMetadata{
 			Command: strings.Join(request.cmd, " "),
 		},
 		KubernetesClusterMetadata: ctx.eventClusterMeta(),
 		KubernetesPodMetadata:     eventPodMeta,
 	}
 
-	if err := f.cfg.StreamEmitter.EmitAuditEvent(f.ctx, event); err != nil {
-		f.log.WithError(err).Warn("Failed to emit exec event.")
-	}
+	defer func() {
+		if err := f.cfg.StreamEmitter.EmitAuditEvent(f.ctx, execEvent); err != nil {
+			f.log.WithError(err).Warn("Failed to emit exec event.")
+		}
```

Comment:
Before the `Exec` event was triggered before executing the code but it was never committed to Audit Log because it failed the validation caused by missing the `Metadata.Code`. This field can have two values for exec: `ExecFailureCode` and `ExecCode`. 
In order for us to know if the execution failed we must trigger it and wait for its result.
This is what this MR does, in order to correctly populate the `Metadata.Code` with the correct value it waits until the execution finishes and commits the `Exec` event with the correct code, as we already do in SSH connections.